proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1378 @@
+
+If a hook modifies the options in these phases, then the configuration is propagated
+into all subsequent phases, and the design of the interface ensures that
+this is the only point where hooks can modify the options.
+
+It is only the pre-configuration hooks which allow modification of the options.
+This is because the configuration process computes some more complicated data
+structures from these initial inputs. If hooks were allowed to modify the results
+of configuration then it would be error-prone to ensure that they suitably updated
+both the options in question as well as the generated configuration.
+For example, both `PackageDescription` and `ComponentLocalBuildInfo` contain
+a list of exposed modules for the library.
+This is why the "post" configuration hooks (and any hooks after that)
+can only run an `IO` action; they can't return any modifications that would affect the `PackageDescription`.
+
+
+#### `LocalBuildConfig`
+
+There are parts of the `LocalBuildInfo` which must be decided at a global (per-package) level.
+For instance, whether to build dynamic libraries.
+On the other hand, there are also things we want to decide on a local (per-component) level,
+such as specific GHC options with which to compile the component.
+
+Moreover, there are parts of the `LocalBuildInfo` which hooks cannot modify.
+For example, things such as package dependencies can't be modified because they are
+determined externally by the overall build plan (e.g. from the dependency solver).
+Thus the hooks interface should prevent the modification of these
+parts of `LocalBuildInfo`.
+
+We propose to achieve this by defining a new type `LocalBuildConfig` which
+contains only the parts of the existing `LocalBuildInfo` datatype that can be
+modified by `preConfPackageHook`.
+
+
+#### `ComponentDiff`
+
+The `ComponentDiff` records the modifications that should be applied to each component.
+
+For each component, `preConfComponentHook` is run, returning a `ComponentDiff`.
+This `ComponentDiff` is applied to its corresponding `Component`
+by monoidally combining together the fields.
+
+```haskell
+newtype ComponentDiff = ComponentDiff { componentDiff :: Component }
+
+emptyComponentDiff :: ComponentName -> ComponentDiff
+```
+
+The diff is represented by a `Component`; not all fields of a `Component` are
+allowed to be modified, and when the diff is applied it is dynamically checked
+that the hook has not modified any fields which it shouldn't.
+
+An alternative design would be to define a custom diff datatype which statically
+distinguishes which fields are allowed to be modified. This would be more
+difficult to maintain, as it requires `Cabal` developers to keep the `Component`
+and `ComponentDiff` types up-to-date.  However if we end up with a design in
+which `Cabal`'s version of the `Component` type is necessarily separate from the
+type in the hooks API, we may want to reconsider this.
+
+
+### Build hooks
```

Comment:
Build hooks seem problematic for another reason: what is the "build phase" in HLS? Building is _continuous_, and is re-done frequently but incrementally.
1. When do we run the pre-build hook? Every time we want to incrementally re-build anything? (this would be significantly improved if we followed my suggestion about making hooks more shake-like, then we could actually re-run them only if needed by the thing we're currently building)
2. When do we run the post-build hook? Arguably the build phase never "finishes" (or we might stop building, but without having built all the modules). So maybe we don't run it at all? Does that make sense?

Perhaps my objection here is broader. If we think about more general cases where our build tool just has a big graph of build rules:
- Pre-phase hooks are kind of okay so long as they have good dependency information. They are rules and we just have to make all the "Phase X" rules depend on their output.
- Post-phase hooks are weird. Normally we work on a demand-driven basis, so when are these demanded? And what do they depend on?

Reply:
This is a very good point regarding post-phase hooks not having clear demand. Perhaps it would be better to say that all hooks should be pre-something.

Reply:
1. A design that is good for HLS / fine-grained world can easily be retrofitted on the course-grain world by simply running all the fine-grained hooks once as part of some larger step.

2. A design that is good for current Cabal / course-grained world will be horrible for the fine-grained world.

If are are trying to make something that is nice for HLS from the get-go, we should be much more aggressive. If we are trying to make something that is easy to mindlessly batch convert existing `Setup.hs` we should be clear it is wholly inadiquate for the brave new HLS world and just an intermediary stop-gap.

Reply:
Coming back to this comment, I think I was too hard on post-build hooks. If we think of things from the rule perspective, then the build phase runs many build actions, one for each module. It then has a "phony" rule for the "build phase" that depends on all of these.

HLS very much _does_ have the per-module build rules, but it _doesn't_ have a "complete the phase" rule. It would be sensical for HLS to have post-build hooks that run after building a particular module, it just doesn't make so much sense to have hooks that run after "completing the phase".